### PR TITLE
add fireredasr model

### PR DESCRIPTION
**Pull request by sxl1993**
## Purpose
This PR adds vLLM integration for FireRedASR, an LLM-based automatic speech recognition (ASR) model. FireRedASR uses a speech encoder + projector + Qwen2 LLM architecture for end-to-end speech-to-text transcription.

  The implementation includes:

  1. Model architecture integration: vllm/model_executor/models/fireredasr.py with 1736 lines implementing:
    - Audio feature extractor (ASRFeatExtractor, CMVN, KaldifeatFbank)
    - 16-layer Conformer encoder for speech processing
    - Transformer decoder for text generation
    - Full vLLM multimodal interface support
  2. Configuration support: vllm/transformers_utils/configs/fireredasr.py with:
    - FireRedAsrEncoderConfig for audio encoder configuration
    - FireRedAsrConfig main model config integrating audio encoder and Qwen2 text config
  3. Registry updates: Added FireRedASR to model registry and configuration registry:
    - vllm/model_executor/models/registry.py: Added FireRedAsrForConditionalGeneration to _TEXT_GENERATION_MODELS
    - vllm/transformers_utils/configs/__init__.py: Added config mapping
    - vllm/transformers_utils/config.py: Added fireredasr to _CONFIG_REGISTRY

## Test Plan
Unit Tests (To be added in follow-up PRs):

  1. Model loading test: Verify FireRedASR model can be loaded correctly with vLLM
  2. Forward pass test: Ensure model forward pass works with dummy audio inputs
  3. Configuration test: Test config parsing and validation
  4. Multimodal integration test: Verify audio input processing works with vLLM multimodal pipeline

  Integration Tests:

  1. End-to-end inference test:
  # Test with sample audio file
  vllm serve \
    /ossfs/workspace/models/FireRedASR-LLM-L \
    --tokenizer /ossfs/workspace/models/FireRedASR-LLM-L/modified_tokenizer \
    --served-model-name firered-asr \
    --dtype bfloat16 \
    --max-model-len 8192 \
    --gpu-memory-utilization 0.8 \
    --max-num-seqs 64 \
    --allowed-local-media-path / \
    --port 8902 \
    --enforce-eager
    
    curl http://localhost:8902/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
  "messages": [
    {
      "role": "user",
      "content": [
        {
          "type": "audio_url",
          "audio_url": {
            "url": "https:xxxx"
          }
        },
        {
          "type": "text",
          "text": "请转写音频为文字"
        }
      ]
    }
  ]
}'

  2. Batch processing test: Test with multiple audio inputs
  3. Performance benchmark: Compare transcription accuracy and latency with baseline ASR models
## Test Result
 Initial Validation:

  - ✅ Model architecture successfully compiles and integrates with vLLM framework
  - ✅ Configuration system correctly parses FireRedASR configs
  - ✅ Model registry recognizes FireRedAsrForConditionalGeneration architecture
  - ✅ Basic multimodal audio input support implemented
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by sxl1993** [comment]
> ## Purpose
> This PR adds vLLM integration for FireRedASR, an LLM-based automatic speech recognition (ASR) model. FireRedASR uses a speech encoder + projector + Qwen2 LLM architecture for end-to-end speech-to-text transcription.
> 
> The implementation includes:
> 
> 1. Model architecture integration: vllm/model_executor/models/fireredasr.py with 1736 lines implementing:
>    - Audio feature extractor (ASRFeatExtractor, CMVN, KaldifeatFbank)
>    - 16-layer Conformer encoder for speech processing
>    - Transformer decoder for text generation
>    - Full vLLM multimodal interface support
> 2. Configuration support: vllm/transformers_utils/configs/fireredasr.py with:
>    - FireRedAsrEncoderConfig for audio encoder configuration
>    - FireRedAsrConfig main model config integrating audio encoder and Qwen2 text config
> 3. Registry updates: Added FireRedASR to model registry and configuration registry:
>    - vllm/model_executor/models/registry.py: Added FireRedAsrForConditionalGeneration to _TEXT_GENERATION_MODELS
>    - vllm/transformers_utils/configs/**init**.py: Added config mapping
>    - vllm/transformers_utils/config.py: Added fireredasr to _CONFIG_REGISTRY
> 
> ## Test Plan
> Unit Tests (To be added in follow-up PRs):
> 
> 1. Model loading test: Verify FireRedASR model can be loaded correctly with vLLM
> 2. Forward pass test: Ensure model forward pass works with dummy audio inputs
> 3. Configuration test: Test config parsing and validation
> 4. Multimodal integration test: Verify audio input processing works with vLLM multimodal pipeline
> 
> Integration Tests:
> 
> 1. End-to-end inference test:
> 
> # Test with sample audio file
> vllm serve /ossfs/workspace/models/FireRedASR-LLM-L --tokenizer /ossfs/workspace/models/FireRedASR-LLM-L/modified_tokenizer --served-model-name firered-asr --dtype bfloat16 --max-model-len 8192 --gpu-memory-utilization 0.8 --max-num-seqs 64 --allowed-local-media-path / --port 8902 --enforce-eager
> 
> ```
> curl http://localhost:8902/v1/chat/completions \
> -H "Content-Type: application/json" \
> -d '{
> ```
> 
> "messages": [ { "role": "user", "content": [ { "type": "audio_url", "audio_url": { "url": "https:xxxx" } }, { "type": "text", "text": "请转写音频为文字" } ] } ] }'
> 
> 2. Batch processing test: Test with multiple audio inputs
> 3. Performance benchmark: Compare transcription accuracy and latency with baseline ASR models
> 
> ## Test Result
> Initial Validation:
> 
> * ✅ Model architecture successfully compiles and integrates with vLLM framework
> * ✅ Configuration system correctly parses FireRedASR configs
> * ✅ Model registry recognizes FireRedAsrForConditionalGeneration architecture
> * ✅ Basic multimodal audio input support implemented
> 
> Essential Elements of an Effective PR Description Checklist

File: vllm/model_executor/models/fireredasr.py [hunk 1]
```diff
@@ -0,0 +1,1736 @@
...
+    def forward(self, x, pos_emb, slf_attn_mask=None, pad_mask=None):
+        out = 0.5 * x + 0.5 * self.ffn1(x)
+        out = self.mhsa(out, out, out, pos_emb, mask=slf_attn_mask)[0]
+        out = self.conv(out, pad_mask)
+        out = 0.5 * out + 0.5 * self.ffn2(out)
+        out = self.layer_norm(out)
+        return out
+
+
+class Swish(nn.Module):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/fireredasr.py:264]: Replace with `nn.SiLU`

**Review by DarkLight1337 — COMMENTED** [review]
Can you add an example in `examples/offline_inference/audio_language.py` and also the supported models page (`/home/cyrus/vllm/docs/models/supported_models.md`)? For CI to pass, you also need to add the model definition to `tests/models/registry.py`.

File: vllm/model_executor/models/registry.py [hunk 2]
```diff
@@ -101,6 +101,7 @@
     "FalconMambaForCausalLM": ("mamba", "MambaForCausalLM"),
     "FalconH1ForCausalLM": ("falcon_h1", "FalconH1ForCausalLM"),
     "FlexOlmoForCausalLM": ("flex_olmo", "FlexOlmoForCausalLM"),
+    "FireRedAsrForConditionalGeneration": ("fireredasr", "FireRedAsrForConditionalGeneration"),
```
Comment (DarkLight1337) [inline vllm/model_executor/models/registry.py]: Please keep in alphabetical order

File: vllm/model_executor/models/firered_audio.py [hunk 3]
```diff
@@ -0,0 +1,1736 @@
...
+
+This module implements the vLLM integration for FireRedASR's LLM-based ASR model,
+which uses a speech encoder + projector + Qwen2 LLM architecture.
+"""
+import os
+from collections.abc import Callable, Iterable, Mapping, Sequence
+from typing import Annotated, Any, List, Literal, Optional, Tuple, TypedDict, Union
+
+import kaldi_native_fbank as knf
+import kaldiio
```
Comment (Isotr0py) [inline vllm/model_executor/models/firered_audio.py:12]: I think `kaldi` is not a dependency in vLLM, can you lazy import it like timm?
https://github.com/vllm-project/vllm/blob/2b99f210f53faee71fd56cf52036040847e6b420/vllm/model_executor/models/minicpmv.py#L1233-L1236

File: vllm/model_executor/models/firered_audio.py [hunk 4]
```diff
@@ -0,0 +1,1736 @@
...
+        k = self.w_ks(k).view(bs, -1, self.n_head, self.d_k)
+        v = self.w_vs(v).view(bs, -1, self.n_head, self.d_k)
+        q = q.transpose(1, 2)
+        k = k.transpose(1, 2)
+        v = v.transpose(1, 2)
+
+        if mask is not None:
+            mask = mask.unsqueeze(1)
+
+        output = self.attention(q, k, v, mask=mask)
```
Comment (Isotr0py) [inline vllm/model_executor/models/firered_audio.py]: Can we use `F.scaled_dot_product_attention` here?
